crates/ruff_linter/src/rules/flake8_bugbear/rules/loop_iterator_mutated.rs

```diff
@@ -0,0 +1,185 @@
+use ruff_diagnostics::Violation;
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{
+    visitor::{self, Visitor},
+    Expr, ExprAttribute, ExprCall, ExprContext, ExprName, ExprSubscript, Stmt, StmtDelete, StmtFor,
+};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+use ruff_diagnostics::Diagnostic;
+
+static MUTATING_FUNCTIONS: &'static [&'static str] = &[
+    "append", "sort", "reverse", "remove", "clear", "extend", "insert", "pop", "popitem",
+];
+
+/// ## What it does
+/// Checks for mutation of the iterator of a loop in the loop's body
+///
+/// ## Why is this bad?
+/// Changing the structure that is being iterated over will usually lead to
+/// unintended behavior as not all elements will be addressed.
+///
+/// ## Example
+/// ```python
+/// some_list = [1,2,3]
+/// for i in some_list:
+///   some_list.remove(i) # this will lead to not all elements being printed
+///   print(i)
+/// ```
+///
+///
+/// ## References
+/// - [Python documentation: Mutable Sequence Types](https://docs.python.org/3/library/stdtypes.html#typesseq-mutable)
+#[violation]
+pub struct LoopIteratorMutated;
+
+impl Violation for LoopIteratorMutated {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("editing a loop's mutable iterable often leads to unexpected results/bugs")
+    }
+}
+
+fn _to_name_str(node: &Expr) -> String {
+    match node {
+        Expr::Name(ExprName { id, .. }) => {
+            return id.to_string();
+        }
+        Expr::Attribute(ExprAttribute {
+            range: _,
+            value,
+            attr,
+            ..
+        }) => {
+            let mut inner = _to_name_str(value);
+            match inner.as_str() {
+                "" => {
+                    return "".into();
+                }
+                _ => {
+                    inner.push_str(".");
+                    inner.push_str(attr);
+                    return inner;
+                }
+            }
+        }
+        Expr::Call(ExprCall { range: _, func, .. }) => {
+            return _to_name_str(func);
```

Comment:
Does this mean that `foo().bar` and `foo.bar` would be considered identical?

Reply:
I tried out `collect_call_path` and `compose_call_path` as you've suggested below, and both of them return `None` for `a().foo`.

Not 100% sure how I should go about this - should this be the case?

Also wondering, would you say the following should be caught?:

```py
for bar in a().foo:
  del a().foo
```

Reply:
I would say that should not be caught, since we can't know that `a()` returns the same object.

Reply:
ok, then I think I can use `compose_call_path` - I was just worrying that it returns `None` whenever there is `a()` in the call path.